Space charge Mayes IPAC2018 benchmark

cemitch99 opened the pull request:
This PR refers to:

C. E. Mayes, R. D. Ryne, and D. C. Sagan, "3D Space Charge in BMAD," in proc. IPAC2018,
[DOI:10.18429/JACoW-IPAC2018-THPAK085](https://doi.org/10.18429/JACoW-IPAC2018-THPAK085)

This is a replacement for PR #257 
Closes @ChristopherMayes' #256.

Added the expanding Gaussian beam test and plot script corresponding to phase space shown in Fig. 4.  

- [x] add expanding Gaussian beam test
- [x] add plot script to reproduce Fig. 4
- [x] add static Gaussian beam field calculation
- [x] add plot script to reproduce Fig. 1
- [x] add documentation
- [ ] verify resolution and convergence
- [ ] potentially use IGF solver (with much smaller grid) #605 

![Figure_1](https://github.com/ECP-WarpX/impactx/assets/1353258/55c9d2fb-5e9b-4863-a6e7-48e012ea330e)

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Add expanding test from Mayes IPAC2018

cemitch99 committed:
Delete examples/kurth/input_kurth_10nC.in

Not part of this PR

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on examples/ipac2018_mayes/input_expanding_ipac2018.py:
```diff
@@ -0,0 +1,38 @@
+###############################################################################
+# Beamline: lattice elements and segments
+###############################################################################
+lattice.elements = monitor drift1 monitor
```
I think this file has the wrong file ending :)

cemitch99 committed:
Correct file suffix.

cemitch99 committed:
Delete examples/ipac2018_mayes/input_expanding_ipac2018.py

Removed file with incorrect suffix.

cemitch99 committed:
Adding field line-out prediction data.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add Gnuplot scripts for Mayes Fig. 1

ax3l committed:
Use latest pyAMReX

ax3l committed:
Plot Space-Charge Field

looks correct up to a unit conversion factor

ax3l committed:
Merge remote-tracking branch 'mainline/development' into mayes_benchmark

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ChristopherMayes commented:
Could the plot(s) be created and checked in, or at least put in docs somewhere?

cemitch99 committed:
Correct Reference Frame and + Charge

ax3l committed:
MLMG: Expose Solver Parameters

Expose the numeric parameters of the multi-grid solver
that we currently use for space charge calculations.

ax3l committed:
Improve Doc Strings: Ref Kin Energy

ax3l committed:
Plot multiple r runs together

ax3l committed:
r=0.2,0.5,1,10

ax3l commented:
@ChristopherMayes Thank you for your comment. Yes, absolutely. We still finalize a few space charge benchmarks for HB 2023 right now and this is one of them. As stated in your IPAC2018 paper, the non-IGF method starts to fail for large aspect ratios (w/o pushing resolution much further up). We added a preview to the PR description.

We will clean this PR up further once final, e.g., replacing the data files from the analytical solution with results calculated on the fly.

Yes, on the plots, too. The docs will get a readme file for the new case and we will include the plot, similar to https://impactx.readthedocs.io/en/latest/usage/examples/fodo/README.html

ax3l commented:
@ChristopherMayes commented that if we also plot Fig 2 of https://accelconf.web.cern.ch/ipac2018/papers/thpak085.pdf, we can see more precision details (e.g., for fixed precision and infinite number of steps in the MLMG, do different aspect ratios converge to the same result). For instance, GPT sometimes gets this wrong for some aspect ratios.

cemitch99 committed:
Add README doc for examples.

cemitch99 committed:
Merge branch 'development' into mayes_benchmark

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update input_expanding_ipac2018.in

Update "beam.energy" -> "beam.kin_energy"

cemitch99 committed:
Update run_expanding_ipac2018.py

Update "energy" -> "kin_energy"

cemitch99 committed:
Update run_fieldplots_ipac2018.py

Update "energy" -> "kin_energy"

cemitch99 committed:
Update run_expanding_ipac2018.py

Update kin_energy value.

cemitch99 committed:
Update run_expanding_ipac2018.py

Update sim.prob_relative syntax.

cemitch99 committed:
Update run_fieldplots_ipac2018.py

Update sim.prob_relative syntax.

cemitch99 committed:
Update run_fieldplots_ipac2018.py

Comment unused parameters.

cemitch99 committed:
Resolve merge conflicts.

cemitch99 commented on examples/CMakeLists.txt:
```diff
@@ -699,6 +699,7 @@ add_impactx_test(hvkicker_madx.py
         OFF  # no plot script yet
 )
 
+<<<<<<< HEAD
```
```suggestion
```

cemitch99 committed:
Update examples/CMakeLists.txt

cemitch99 committed:
Update run_fieldplots_ipac2018.py

del.sim -> sim.finalize()

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Add access to required field data files.

cemitch99 committed:
Update CMakeLists.txt

Correct path.

cemitch99 committed:
Update CMakeLists.txt

Correct destination of copy.

cemitch99 committed:
Update CMakeLists.txt

Comment field test (debug).

cemitch99 committed:
Add analysis scripts (draft).

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update analysis_expanding_ipac2018.py

Use large tolerance for now.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update analysis_fieldplots_ipac2018.py

Use large tolerance for now.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update run_expanding_ipac2018.py

Remove unnecessary plot lines (use plot_expanding_ipac2018.py instead).

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Turn on second test.

cemitch99 committed:
Update CMakeLists.txt

Comment call to Ey for testing.

cemitch99 committed:
Update CMakeLists.txt

Correct Ey -> Ez

cemitch99 committed:
Update run_fieldplots_ipac2018.py

nGrowVect -> n_grow_vect

cemitch99 committed:
Update CMakeLists.txt

Turn off analysis for field plot test.

cemitch99 committed:
Merge branch 'development' into mayes_benchmark

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on examples/ipac2018_mayes/run_fieldplots_ipac2018.py:
```diff
@@ -0,0 +1,190 @@
+    p = ax.plot(x_theory, theory[case], label=case)  # E_x(sigma_r) [MV/m]
+    line_color = p[0].get_color()
+
+    # plot data slices
+    ng = F_x.n_grow_vect
+    # q_e = -1.602176634e-19   # particle charge (not used)
+    for mfi in F_x:
```
We can now, when performance is not the issue, use a cool new global indexing feature to access all boxes, even (MPI-wise) remote ones:
https://pyamrex.readthedocs.io/en/latest/usage/compute.html
<img width="731" height="840" alt="image" src="https://github.com/user-attachments/assets/e1070397-6199-4e6f-a0c1-13ed97e87b16" />

So we can do things now like:
```py
shape = F_x.shape()
F_x_slice_xy = F_x[:,:,shape[2]//2]

plt.imshow(F_x_slice_xy)
plt.show()
```

cemitch99 committed:
Merge branch 'development' into mayes_benchmark

ax3l commented:
@cemitch99 shall we finalize and merge this one? :)
